### PR TITLE
performance tracker refactoring

### PR DESCRIPTION
**Pull request by tbenr**
Removed all the `join()` on safe futures to make performance tracker fully async.

I put back the join in onSlot() since it avoid overlapping executions and blocking onSlot() won't block other onSlot() to run

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 1]
```diff
@@ -240,130 +240,154 @@ private boolean isInHistoricBlockRoots(
...
+            inclusionDistanceStatistics.getAverage(),
+            correctTargetCount,
+            correctHeadBlockCount)
+        : AttestationPerformance.empty(
+            analyzedEpoch, validatorTracker.getNumberOfValidatorsForEpoch(analyzedEpoch));
   }
 
-  private Map<UInt64, List<Attestation>> getAttestationsIncludedInEpochs(
+  private SafeFuture<Set<BeaconBlock>> getBlocksInEpochs(
       UInt64 startEpochInclusive, UInt64 endEpochExclusive) {
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java:345]: finals

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 2]
```diff
@@ -240,130 +240,154 @@ private boolean isInHistoricBlockRoots(
...
+              maybeBlock ->
+                  maybeBlock.map(block -> !block.getSlot().equals(UInt64.ZERO)).orElse(false));
+
+    } else {
+      return SafeFuture.completedFuture(false);
+    }
+  }
+
+  private SafeFuture<Map<UInt64, List<Attestation>>> getAttestationsIncludedInEpochs(
+      UInt64 startEpochInclusive, UInt64 endEpochExclusive) {
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java:382]: finals

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 3]
```diff
@@ -240,130 +240,154 @@ private boolean isInHistoricBlockRoots(
...
+        spec.computeStartSlotAtEpoch(endEpochExclusive).decrement();
+
+    final Set<BeaconBlock> blocksInEpoch = new HashSet<>();
+    final AtomicReference<UInt64> currSlot = new AtomicReference<>(inclusiveEndEpochEndSlot);
+    return SafeFuture.asyncDoWhile(
+            () -> getBlockInEffectAtSlot(blocksInEpoch, epochStartSlot, currSlot))
+        .thenApply(__ -> blocksInEpoch);
+  }
+
+  private SafeFuture<Boolean> getBlockInEffectAtSlot(
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: Get assumes we return it. Maybe put/set/fill could be better? Not sure

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 4]
```diff
@@ -240,130 +240,154 @@ private boolean isInHistoricBlockRoots(
...
-      currSlot = block.get().getSlot().decrement();
     }
-    return blocksInEpoch;
+
+    final IntSummaryStatistics inclusionDistanceStatistics =
+        inclusionDistances.intStream().summaryStatistics();
+
+    // IntSummaryStatistics returns Integer.MIN and MAX when the summarized integer list
+    // is empty.
+    int numberOfProducedAttestations = producedAttestations.size();
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: final

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 5]
```diff
@@ -240,130 +240,154 @@ private boolean isInHistoricBlockRoots(
...
+    int correctTargetCount = 0;
+    int correctHeadBlockCount = 0;
+
+    // Pre-process attestations included on chain to group them by
+    // data hash to inclusion slot to aggregation bitlist
+    final Map<Bytes32, NavigableMap<UInt64, SszBitlist>> slotAndBitlistsByAttestationDataHash =
+        new HashMap<>();
+    for (Map.Entry<UInt64, List<Attestation>> entry : attestationsIncludedOnChain.entrySet()) {
+      for (Attestation attestation : entry.getValue()) {
+        Bytes32 attestationDataHash = attestation.getData().hashTreeRoot();
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: final

File: beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java [hunk 6]
```diff
@@ -240,130 +240,154 @@ private boolean isInHistoricBlockRoots(
...
+        NavigableMap<UInt64, SszBitlist> slotToBitlists =
```
Comment (zilm13) [inline beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java]: final

**Review by zilm13 — APPROVED** [review]
LGTM 
Maybe I'm to picky on finals, but just for consistency

**Comment by tbenr** [comment]
putting it back to draft since it isn't actually needed

**Review by StefanBratanov — APPROVED** [review]
LGTM

**Pull request merged**